graph/graphs/gen/gen.go

```diff
@@ -175,6 +175,33 @@ func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) {
 	cycleNoCheck(dst, cycle)
 }
 
+// Tree constructs an n-ary tree breadth-first in dst with the given fan-out, n.
+// If the number of nodes does not equal \sum_{i=0}^h n^i, where h is an integer
+// indicating the height of the tree, a partial tree will be constructed with not
+// all nodes having zero or n children, and not all leaves being h from the root.
+// n must be greater than zero and less than the number of nodes, otherwise
+// Tree will panic. If dst is a directed graph, edges are directed from the root
+// to the leaves. If any ID appears more than once in nodes, Tree will panic.
+func Tree(dst NodeIDGraphBuilder, n int, nodes IDer) {
+	if n < 1 || nodes.Len() <= n {
+		panic("gen: invalid fan-out")
+	}
+
+	err := check(nodes)
+	if err != nil {
+		panic(err)
+	}
+
+	j := 0
+	for i := 0; j < nodes.Len(); i++ {
+		u, _ := dst.NodeWithID(nodes.ID(i))
```

Comment:
I think this is not right for the case of n=1, |nodes|=1; there needs to be a special case for that so a singleton tree is actually added. Sorry I missed that before.

Reply:
Would a check of length of nodes at the top of `Tree` be appropriate? This check would only occur once versus the check that would be necessary for adding a node before setting each edge. Something like the following code:
```
func Tree(dst NodeIDGraphBuilder, n int, nodes IDer) {
	if nodes.Len() == 1 {
		if u, new := dst.NodeWithID(nodes.ID(0)); new {
			dst.AddNode(u)
		}
		return
	}
.
.
.
```

Reply:
Yep, but also with a fast return for no nodes since we're here.

```
switch nodes.Len() {
case 0:
	return
case 1:
	if u, new := dst.NodeWithID(nodes.ID(0)); new {
		dst.AddNode(u)
	}
	return
}
```

Reply:
Done.

Reply:
And so it goes... We need an update of the docs now (sorry). We need to say that as special cases, |nodes| == 0 is a no-op and |nodes| == 1 will result a singleton graph (if previously empty). Currently the docs say that both of these will panic. Something like:

```
// If the number of nodes is greater than one, n must be non-zero and
// less than the number of nodes unless, otherwise Tree will panic.
```

Reply:
Not a problem. It runs a little long, but how does this look:

```
// Tree constructs an n-ary tree breadth-first in dst with the given fan-out, n.
// If the number of nodes does not equal \sum_{i=0}^h n^i, where h is an integer
// indicating the height of the tree, a partial tree will be constructed with not
// all nodes having zero or n children, and not all leaves being h from the root.
// If the number of nodes is zero, Tree will do nothing.
// If the number of nodes is one, Tree will only add that node to dst.
// If the number of nodes is greater than one, n must be non-zero and less than the
// number of nodes, otherwise Tree will panic.
// If dst is a directed graph, edges are directed from the root to the
// leaves. If any ID appears more than once in nodes, Tree will panic.
```

Reply:
I think that either bullet point special cases (like in stdlib's math) or just the two lines I have above (in terms of detail - if not exact wording - the special cases are only to avoid confusion with the panic, they do what you would expect so I don't think they need to be fully spelled out).

Reply:
Done.